File: server/contextos/models/agent.py
from enum import Enum
from datetime import datetime, timezone
from typing import Optional
from pydantic import BaseModel, Field

class AgentRole(str, Enum):
    PLANNER = "PLANNER"
    CODER = "CODER"
    REVIEWER = "REVIEWER"
    RESEARCHER = "RESEARCHER"
    TESTER = "TESTER"

class AgentStatus(str, Enum):
    IDLE = "IDLE"
    STARTING = "STARTING"
    RUNNING = "RUNNING"
    WAITING_APPROVAL = "WAITING_APPROVAL"
    PAUSED = "PAUSED"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    STOPPED = "STOPPED"

class AgentPermissions(BaseModel):
    can_execute_shell: bool = False
    can_edit_files: bool = True
    can_git_commit: bool = False
    can_network: bool = False

class TokenUsage(BaseModel):
    candidate_tokens: int = 0
    selected_tokens: int = 0
    excluded_tokens: int = 0
    estimated_tokens_avoided: int = 0
    cache_hits: int = 0
    total_input_tokens: int = 0
    total_output_tokens: int = 0
    reused_cached_tokens: int = 0

class Agent(BaseModel):
    id: str
    name: str
    role: AgentRole
    provider: str
    model: str
    status: AgentStatus = AgentStatus.IDLE
    current_task: str = ""
    project_id: str = "contextos"
    session_id: str = "default_session"
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    started_at: Optional[datetime] = None
    last_activity_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    context_budget: int = 8000
    token_usage: TokenUsage = Field(default_factory=TokenUsage)
    permissions: AgentPermissions = Field(default_factory=AgentPermissions)
    runtime_seconds: int = 0

File: web/src/types/agent.ts
export type AgentRole = 'PLANNER' | 'CODER' | 'REVIEWER' | 'RESEARCHER' | 'TESTER';

export type AgentStatus =
  | 'IDLE'
  | 'STARTING'
  | 'RUNNING'
  | 'WAITING_APPROVAL'
  | 'PAUSED'
  | 'COMPLETED'
  | 'FAILED'
  | 'STOPPED';

export interface TokenMetrics {
  candidate_tokens: number;
  selected_tokens: number;
  excluded_tokens: number;
  estimated_tokens_avoided: number;
  cache_hits: number;
  total_input_tokens: number;
  total_output_tokens: number;
  reused_cached_tokens?: number;
  context_build_latency_ms?: number;
}

export interface Agent {
  id: string;
  name: string;
  type: AgentRole;
  provider: string;
  model: string;
  status: AgentStatus;
  current_task: string;
  tokens: TokenMetrics;
  context_size: number;
  runtime_seconds: number;
  project_id: string;
  session_id: string;
  created_at: string;
  started_at?: string;
  last_activity_at: string;
  permissions: {
    can_execute_shell: boolean;
    can_edit_files: boolean;
    can_git_commit: boolean;
    can_network: boolean;
  };
}

export type EventType =
  | 'agent.started'
  | 'agent.thinking'
  | 'message.created'
  | 'tool.requested'
  | 'tool.approval_required'
  | 'tool.started'
  | 'tool.completed'
  | 'file.read'
  | 'file.changed'
  | 'terminal.started'
  | 'terminal.output'
  | 'memory.read'
  | 'memory.write'
  | 'context.generated'
  | 'agent.paused'
  | 'agent.completed'
  | 'agent.failed';

export interface AgentEvent {
  id: string;
  agent_id: string;
  session_id: string;
  type: EventType;
  timestamp: string;
  payload: Record<string, any>;
}

export interface Message {
  id: string;
  agent_id: string;
  role: 'user' | 'assistant' | 'system';
  content: string;
  timestamp: string;
  tool_calls?: {
    id: string;
    name: string;
    arguments: Record<string, any>;
    status: 'pending' | 'requires_approval' | 'executed' | 'rejected';
    result?: any;
  }[];
}
